### PR TITLE
Jfontanilla/graph endpoints

### PR DESCRIPTION
**Pull request by fontan10**
endpoints for frontend to add, remove, and query vertex and edge info

no endpoints for updating vertex positions yet

remind me to read out and explain the sticky note I have on my laptop in person. I'm too lazy to write it out fully and properly and its easier to explain in person. its about the code architecture and separation of concerns

### CODE REVIEW

File: MergeSharpWebAPI/Services/GraphService.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        if (_keyToVertexMap.TryGetValue(key, out Graph.Vertex v))
+        {
+            return new VertexInfo(key, v.x, v.y, v.type);
+        }
+
+        // TODO: throw an exception
+        return new VertexInfo();
+    }
+
+    public bool AddVertex(int key, double x, double y, string stype)
```
Comment (ayshon) [inline MergeSharpWebAPI/Services/GraphService.cs:76]: just curious why you named the variable stype, is this a naming convention 0.0

File: MergeSharpWebAPI/Services/GraphService.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    {
+        if (_keyToVertexMap.ContainsKey(key))
+        {
+            return false;
+        }
+
+        if (Enum.TryParse(stype, true, out Graph.Vertex.Type type))
```
Comment (ayshon) [inline MergeSharpWebAPI/Services/GraphService.cs:83]: insane learning another new c# function

File: MergeSharpWebAPI/Services/GraphService.cs (hunk 1 above)
Comment (fontan10) [reply MergeSharpWebAPI/Services/GraphService.cs:76]: Nah just s for string cause there's the string type and then the Type type

File: MergeSharpWebAPI/Services/GraphService.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            Guid vertexGuid = Guid.NewGuid();
+            var v = new Graph.Vertex(vertexGuid, x, y, type);
+
+            _vertexGuidToKeyMap.Add(v.guid, key);
+            _keyToVertexMap.Add(key, v);
+
+            return _graph.AddVertex(v);
```
Comment (ayshon) [inline MergeSharpWebAPI/Services/GraphService.cs]: is there ever a time this returns false? and if there is do we still want to add the vertex to the mapping?? does this question even matter???

File: MergeSharpWebAPI/Services/GraphService.cs (hunk 1 above)
Comment (ayshon) [reply MergeSharpWebAPI/Services/GraphService.cs:76]: Gotchu makes sense

**Review by ayshon — APPROVED** [review]
Looks good for the most part, I left a question about the AddVertex function in GraphService though, I'm not sure if there's an issue there.

File: MergeSharpWebAPI/Controllers/GraphController.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+using Microsoft.AspNetCore.SignalR.Client;
+
+namespace MergeSharpWebAPI.Controllers;
+
+
+// https://stackoverflow.com/questions/12072277/reading-fromuri-and-frombody-at-the-same-time
+// https://www.strathweb.com/2013/04/asp-net-web-api-parameter-binding-part-1-understanding-binding-from-uri/
+[ApiController]
+[Route("[controller]")]
+public class GraphController : ControllerBase
```
Comment (ayshon) [inline MergeSharpWebAPI/Controllers/GraphController.cs:16]: O I just realized this controller is gonna need a SendMessage endpoint for sending state to the frontend, when backend receives new state from server.

Comment (fontan10) [reply MergeSharpWebAPI/Controllers/GraphController.cs:16]: Yup I'll do that in a later pr. Should have been clear about that in this pr's description

File: MergeSharpWebAPI/Services/GraphService.cs (hunk 3 above)
Comment (fontan10) [reply MergeSharpWebAPI/Services/GraphService.cs]: Shoot you're right. It'll be false if there's a conflict in the guid. The chances of that are very low, but I should still have an if statement for the case. I'll address this. 

Nice catch!

**Pull request merged**